consensus/state.go

```diff
@@ -745,11 +753,45 @@ func (cs *ConsensusState) enterNewRound(height int, round int) {
 
 	types.FireEventNewRound(cs.evsw, cs.RoundStateEvent())
 
-	// Immediately go to enterPropose.
-	cs.enterPropose(height, round)
+	// Wait for txs to be available in the mempool
+	// before we enterPropose
+	if cs.config.NoEmptyBlocks {
+		go cs.waitForTxs(height, round)
+	} else {
+		cs.enterPropose(height, round)
+	}
+}
+
+func (cs *ConsensusState) waitForTxs(height, round int) {
+	// if we're the proposer, start a heartbeat routine
+	// to tell other peers we're just waiting for txs (for debugging)
+	done := make(chan struct{})
+	defer close(done)
+	if cs.isProposer() {
+		go cs.proposerHeartbeat(done)
+	}
+
+	// wait for the mempool to have some txs
+	<-cs.mempool.TxsAvailable()
+
+	// now we can enterPropose
+	cs.txsAvailable <- RoundState{Height: height, Round: round}
+}
+
+func (cs *ConsensusState) proposerHeartbeat(done chan struct{}) {
+	for {
+		select {
+		case <-done:
+			return
+		default:
+			// TODO: broadcast heartbeat
```

Comment:
@jaekwon is it ok that this won't get gossiped? only connected peers will hear it. we'll need new ConsensusMessage types for it